// react-styled-flexboxgrid/dist/styled-flexboxgrid.es.js
import React, { PropTypes } from 'react';
import styled, { css } from 'styled-components';

function createProps(propTypes, props) {
  var newProps = {};

  Object.keys(props).filter(function (key) {
    return ~['children'].indexOf(key) || !propTypes[key];
  }).forEach(function (key) {
    return newProps[key] = props[key];
  });

  return newProps;
}

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];

    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }

  return target;
};

























var taggedTemplateLiteral = function (strings, raw) {
  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
};

var _templateObject$1 = taggedTemplateLiteral(['\n    @media ', ' {\n      ', '\n    }\n  '], ['\n    @media ', ' {\n      ', '\n    }\n  ']);

var THEME_CONF = 'flexboxgrid';
var BASE_CONF = {
  gridSize: 12,
  gutterWidth: 1,
  outerMargin: 2,
  container: {
    sm: 46,
    md: 61,
    lg: 76
  },
  breakpoints: {
    xs: 0,
    sm: 768,
    md: 1024,
    lg: 1200
  }
};

var configCache = [];
var makeCacheId = function makeCacheId(props) {
  return JSON.stringify(props.theme && props.theme[THEME_CONF] || {});
};
var resolveConfig = function resolveConfig(props) {
  var themeConf = props.theme && props.theme[THEME_CONF] || {};

  var conf = _extends({}, BASE_CONF, themeConf, {
    container: _extends({}, BASE_CONF.container, themeConf.container),
    breakpoints: _extends({}, BASE_CONF.breakpoints, themeConf.breakpoints)
  });

  conf.media = Object.keys(conf.breakpoints).reduce(function (media, breakpoint) {
    var breakpointWidth = conf.breakpoints[breakpoint];
    media[breakpoint] = makeMedia('only screen' + (breakpointWidth === 0 ? '' : ' and (min-width: ' + (breakpointWidth + 1) + 'px)'));
    return media;
  }, {});

  return conf;
};

var DIMENSION_NAMES = ['xs', 'sm', 'md', 'lg'];

function config(props) {
  var cacheId = makeCacheId(props);
  if (configCache[0] === cacheId) {
    return configCache[1];
  }

  var conf = resolveConfig(props);

  configCache[0] = cacheId;
  configCache[1] = conf;

  return conf;
}

function makeMedia(media) {
  return function () {
    return css(_templateObject$1, media, css.apply(undefined, arguments));
  };
}

var _templateObject = taggedTemplateLiteral(['\n  margin-right: auto;\n  margin-left: auto;\n\n  ', '\n\n  ', '\n'], ['\n  margin-right: auto;\n  margin-left: auto;\n\n  ', '\n\n  ', '\n']);
var _templateObject2 = taggedTemplateLiteral(['\n    padding-right: ', ';\n    padding-left: ', ';\n  '], ['\n    padding-right: ', ';\n    padding-left: ', ';\n  ']);
var _templateObject3 = taggedTemplateLiteral(['\n    ', '\n  '], ['\n    ', '\n  ']);
var _templateObject4 = taggedTemplateLiteral(['\n        width: ', 'rem;\n      '], ['\n        width: ', 'rem;\n      ']);

var Grid = styled(function (props) {
  return React.createElement(props.tagName || 'div', createProps(Grid.PropTypes, props));
})(_templateObject, function (p) {
  return p.fluid && css(_templateObject2, function (p) {
    return config(p).outerMargin + 'rem';
  }, function (p) {
    return config(p).outerMargin + 'rem';
  });
}, function (p) {
  return !p.fluid && css(_templateObject3, DIMENSION_NAMES.map(function (t) {
    return config(p).container[t] && config(p).media[t](_templateObject4, function (p) {
      return config(p).container[t];
    });
  }));
});

Grid.displayName = 'Grid';

Grid.PropTypes = {
  fluid: PropTypes.bool,
  tagName: PropTypes.string,
  children: PropTypes.node
};

var _templateObject$2 = taggedTemplateLiteral(['\n  box-sizing: border-box;\n  display: flex;\n  flex: 0 1 auto;\n  flex-direction: row;\n  flex-wrap: wrap;\n  margin-right: ', 'rem;\n  margin-left: ', 'rem;\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n'], ['\n  box-sizing: border-box;\n  display: flex;\n  flex: 0 1 auto;\n  flex-direction: row;\n  flex-wrap: wrap;\n  margin-right: ', 'rem;\n  margin-left: ', 'rem;\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n\n  ', '\n']);
var _templateObject2$1 = taggedTemplateLiteral(['\n    justify-content: flex-start;\n    text-align: start;\n  '], ['\n    justify-content: flex-start;\n    text-align: start;\n  ']);
var _templateObject3$1 = taggedTemplateLiteral(['\n    justify-content: center;\n    text-align: center;\n  '], ['\n    justify-content: center;\n    text-align: center;\n  ']);
var _templateObject4$1 = taggedTemplateLiteral(['\n    justify-content: flex-end;\n    text-align: end;\n  '], ['\n    justify-content: flex-end;\n    text-align: end;\n  ']);
var _templateObject5 = taggedTemplateLiteral(['\n    align-items: flex-start;\n  '], ['\n    align-items: flex-start;\n  ']);
var _templateObject6 = taggedTemplateLiteral(['\n    align-items: center;\n  '], ['\n    align-items: center;\n  ']);
var _templateObject7 = taggedTemplateLiteral(['\n    align-items: flex-end;\n  '], ['\n    align-items: flex-end;\n  ']);
var _templateObject8 = taggedTemplateLiteral(['\n    justify-content: space-around;\n  '], ['\n    justify-content: space-around;\n  ']);
var _templateObject9 = taggedTemplateLiteral(['\n    justify-content: space-between;\n  '], ['\n    justify-content: space-between;\n  ']);
var _templateObject10 = taggedTemplateLiteral(['\n    order: -1;\n  '], ['\n    order: -1;\n  ']);
var _templateObject11 = taggedTemplateLiteral(['\n    order: 1;\n  '], ['\n    order: 1;\n  ']);

var ModificatorType = PropTypes.oneOf(DIMENSION_NAMES);

var Row = styled(function (props) {
  return React.createElement(props.tagName || 'div', createProps(Row.PropTypes, props));
})(_templateObject$2, function (p) {
  return config(p).gutterWidth / 2 * -1;
}, function (p) {
  return config(p).gutterWidth / 2 * -1;
}, function (p) {
  return p.reverse && '\n    flex-direction: row-reverse;\n  ';
}, function (p) {
  return p.start && config(p).media[p.start](_templateObject2$1);
}, function (p) {
  return p.center && config(p).media[p.center](_templateObject3$1);
}, function (p) {
  return p.end && config(p).media[p.end](_templateObject4$1);
}, function (p) {
  return p.top && config(p).media[p.top](_templateObject5);
}, function (p) {
  return p.middle && config(p).media[p.middle](_templateObject6);
}, function (p) {
  return p.bottom && config(p).media[p.bottom](_templateObject7);
}, function (p) {
  return p.around && config(p).media[p.around](_templateObject8);
}, function (p) {
  return p.between && config(p).media[p.between](_templateObject9);
}, function (p) {
  return p.first && config(p).media[p.first](_templateObject10);
}, function (p) {
  return p.last && config(p).media[p.last](_templateObject11);
});

Row.displayName = 'Row';

Row.PropTypes = {
  reverse: PropTypes.bool,
  start: ModificatorType,
  center: ModificatorType,
  end: ModificatorType,
  top: ModificatorType,
  middle: ModificatorType,
  bottom: ModificatorType,
  around: ModificatorType,
  between: ModificatorType,
  first: ModificatorType,
  last: ModificatorType,
  tagName: PropTypes.string,
  children: PropTypes.node
};

var _templateObject$3 = taggedTemplateLiteral(['\n  box-sizing: border-box;\n  flex: 0 0 auto;\n  padding-right: ', 'rem;\n  padding-left: ', 'rem;\n\n  ', '\n\n  ', '\n\n  ', '\n'], ['\n  box-sizing: border-box;\n  flex: 0 0 auto;\n  padding-right: ', 'rem;\n  padding-left: ', 'rem;\n\n  ', '\n\n  ', '\n\n  ', '\n']);
var _templateObject2$2 = taggedTemplateLiteral(['', ''], ['', '']);
var _templateObject3$2 = taggedTemplateLiteral(['\n        margin-left: ', '%;\n      '], ['\n        margin-left: ', '%;\n      ']);

var ModificatorType$1 = PropTypes.oneOfType([PropTypes.number, PropTypes.bool]);

var offsetProps = DIMENSION_NAMES.map(function (d) {
  return d + 'Offset';
});
var DimensionPropTypes = DIMENSION_NAMES.reduce(function (propTypes, dimension) {
  propTypes[dimension] = ModificatorType$1;
  propTypes[dimension + 'Offset'] = PropTypes.number;
  return propTypes;
}, {});

var Col = styled(function (props) {
  return React.createElement(props.tagName || 'div', createProps(Col.PropTypes, props));
})(_templateObject$3, function (p) {
  return config(p).gutterWidth / 2;
}, function (p) {
  return config(p).gutterWidth / 2;
}, function (p) {
  return p.reverse && '\n    flex-direction: column-reverse;\n  ';
}, function (p) {
  return Object.keys(p).filter(function (k) {
    return ~DIMENSION_NAMES.indexOf(k);
  }).sort(function (a, b) {
    return DIMENSION_NAMES.indexOf(a) - DIMENSION_NAMES.indexOf(b);
  }).map(function (k) {
    return config(p).media[k](_templateObject2$2, Number.isInteger(p[k])

    // Integer value
    ? '\n        flex-basis: ' + 100 / config(p).gridSize * p[k] + '%;\n        max-width: ' + 100 / config(p).gridSize * p[k] + '%;\n        display: block;\n      '
    // Boolean
    : p[k]
    // Auto-width
    ? '\n          flex-grow: 1;\n          flex-basis: 0;\n          max-width: 100%;\n          display: block;\n        '
    // Hide element
    : 'display: none;');
  });
}, function (p) {
  return Object.keys(p).filter(function (k) {
    return ~offsetProps.indexOf(k);
  }).map(function (k) {
    return config(p).media[k.replace(/Offset$/, '')](_templateObject3$2, 100 / config(p).gridSize * p[k]);
  });
});

Col.displayName = 'Col';

Col.PropTypes = _extends({}, DimensionPropTypes, {
  reverse: PropTypes.bool,
  tagName: PropTypes.string,
  children: PropTypes.node
});

export { Grid, Row, Col, BASE_CONF };
